### ISSUE TITLE
Comments on how to improve the API

### ISSUE THREAD

**Issue by kalhauge** [simonmichael/hledger#609]
I have been using `ledger` for a year now and recently changed to `hledger` because I 
wanted to write my own extensions. The last couple of days I did!

Let me first say that I love the work you have done, and for most of the programming it was a breeze.  I  do though have some suggestions about how to improve the API. 

### Use lenses

I first recently discovered lenses myself, and they are gold. I had to write some
lenses to your library before I could get started. 

Here is a little example. Don't get scared by the weird notations:

```haskell
import Control.Lens
import qualified Hledger as HL

jtxns :: Lens' HL.Journal [ HL.Transaction ]
jtxns f s = (\b -> s { HL.jtxns = b }) <$> f (HL.jtxns s)

...

tpostings :: Lens' HL.Transaction [ HL.Posting ]
tpostings f s =  (\b -> s { HL.tpostings = b }) <$> f (HL.tpostings s)

...
```

We can now create a Map of the sum of all accounts:

```haskell
-- A fix that makes the Map a true Monoid
newtype MonoidalMap k a = MonoidalMap { unMonoidalMap :: M.Map k a }

instance (Ord k, Monoid a) => Monoid (MonoidalMap k a) where
  mempty = MonoidalMap M.empty
  mappend (MonoidalMap m1) (MonoidalMap m2) = MonoidalMap (M.unionWith mappend m1 m2)

-- here we go
main = do
   -- Print the sumOf
   mapM print . M.toList . M.map getSum . unMonoidalMap $
     HL.samplejournal
        -- in the sample journal
        ^. jtxns . traverse
        -- ^ traverse all transaction, and foreach of them
        . tpostings . traverse
        -- ^ traverse all postings, and foreach of them
        . to (\posting ->
                 MonoidalMap
                   (M.singleton (posting ^. paccount) (Sum $ posting ^. pamount))
             )
        -- ^ Create a singleton MonoidalMap for each posting,
        -- which `lens` combines to a single map
```

Which produces this output:

```
("assets:bank:checking",0)
("assets:bank:saving",$1.00)
("assets:cash",$-2.00)
("expenses:food",$1.00)
("expenses:supplies",$1.00)
("income:gifts",$-1.00)
("income:salary",$-1.00)
("liabilities:debts",$1.00)
```

Neet right!

We can also use lenses to set values. This piece of code changes all 
the accounts in the transactions with the name `"asserts:bank:checking"`
to `"assets:bank:my-checking"`.  

```haskell
main = do
  let j = HL.samplejournal
          & jtxns . traverse
          . tpostings . traverse
          . paccount . filtered (== "assets:bank:checking")
          .~ "assets:bank:my-checking"
  mapM print $ j ^. jtxns
```

### `show` is serializing

A problem I encountered while working with `hledger-lib` is that `show` actually serializes
instead of showing the values. This makes it really hard to debug problems, as
the real data is not shown. 

Normally, when I program, I have the following distinction between show
and serialize.  Showing is for internal values and debugging. This means that
it should produce "haskell-ish" code. ie:

```haskell
show <haskell> == "<haskell>"
```

Serializing is for the front end, and produce stuff that can be read by the parser.
It also has the benefit of being able to serialize in a context, ei:

```haskell
runParserInCtx ctx . serialize ctx x == Right x
```

Parsing in a context after means that we can render Amounts differently in
different reports, without changing the Amount. 

### Self-referencing data structures

The last thing is the self-referencing data structures. This is often frowned upon
in the functional programming community. And here are some reasons why.

1. It requires that you collect all the data structures in the Data.Types file, 
   which makes the API harder to read. Having the Data close to the function that
uses them often makes a lot of sense. 
2. Updating structures breaks the type-safety. One example is to set the date of a 
    transaction. The date of this transaction is different than the one of its postings parent transaction. This requires us to re-tie the knot after each change. This re-tieing is also needed if new transactions are added.  

It is therefore often much better not to do it unless you *really* need it.

**Comment by simonmichael** [comment]
Hi Christian! 

I forgot I had this sitting in draft folder. 


On Sep 3, 2017, at 6:43 PM, Christian Kalhauge <notifications@github.com <mailto:notifications@github.com>> wrote:
> I have been using ledger for a year now and recently changed to hledger because I
> wanted to write my own extensions. The last couple of days I did!
> 
> Let me first say that I love the work you have done, and for most of the programming it was a breeze. I do though have some suggestions about how to improve the API.
> 
Excellent, I'm glad you had success. Any scripts you could share as examples ? 

Thanks for this great feedback. Usage and improvement of the API is most welcome. 

All your points are well taken and could be an issue or mail list discussion of their own. In each case there are questions to be resolved: 

> Use lenses
> 
I agree it would be nice to have the power of lenses available. At least for experienced haskellers.

How do we introduce lenses to the project ? WIthout making the code unapproachable for new contributors ? Can they be optional ?

Are you happy with the naming, maintainability etc. of your current setup ? Do you have lots of manual lenses to share ?

hledger-ui uses microlens (a little) which has lighter dependencies and provides most of lens' features. Would this be sufficient ? (Though, we currently also depend on lens transitively.)

> show is serializing
> 
> A problem I encountered while working with hledger-lib is that show actually serializes
> instead of showing the values. This makes it really hard to debug problems, as
> the real data is not shown.
> 
> Normally, when I program, I have the following distinction between show
> and serialize. Showing is for internal values and debugging. This means that
> it should produce "haskell-ish" code. ie:
> 
> show <haskell> == "<haskell>"
> Serializing is for the front end, and produce stuff that can be read by the parser.
> It also has the benefit of being able to serialize in a context, ei:
> 
> runParserInCtx ctx . serialize ctx x == Right x
> Parsing in a context after means that we can render Amounts differently in
> different reports, without changing the Amount.
> 

I find the situation and terminology here a bit murky. There is probably a lot of inconsistency. 

In many cases standard show instances were too verbose for easy debugging and I customised them.

More recently I adopted pretty-show for more readable layout. pretty-show requires standard show output to do its thing, so in some places standard-ish show output has been restored. Or pseudo-standard show output has been faked.

We haven't yet had a clear enough overview and policy to really clean it up.

> Self-referencing data structures
> 
> The last thing is the self-referencing data structures. This is often frowned upon
> in the functional programming community. And here are some reasons why.
> 
> It requires that you collect all the data structures in the Data.Types file,
> which makes the API harder to read. Having the Data close to the function that
> uses them often makes a lot of sense.
> Updating structures breaks the type-safety. One example is to set the date of a
> transaction. The date of this transaction is different than the one of its postings parent transaction. This requires us to re-tie the knot after each change. This re-tieing is also needed if new transactions are added.
> It is therefore often much better not to do it unless you really need it.
> 

When I first started coding I tried to define major types and the functions that worked with them in their own modules, which made sense to me coming from OOP style.

I think what forced me to gather most types into one file was mutually referencing files. Because some data structures, or their declarations, reference each other. I think this will be pretty hard to avoid and I invite you to try it.

I *think* the Transaction<->Posting knot-tying came later and is a different kind of mutual reference (in value space rather than module/type space ?). This nicely solves a problem I was having (how to look upward in the account hierarchy to generate Ledger's compressed-tree-style balance reports). I agree that this has some problems; it's easy to forget to maintain the knot and accidentally introduce bugs. I'd be interested to see an alternative demonstrated.

Thanks for the comments,
-Simon

**Comment by kalhauge** [comment]
Hi Simon,

Thanks for looking at my suggestions. 

I'm currently traveling the next couple of weeks! But I'll send you something more concrete when things are more stable. 

> Excellent, I'm glad you had success. Any scripts you could share as examples?

It's a work in progress and as soon as it reaches a certain quality I will publish it. But for now, it's rather incomplete. 

> How do we introduce lenses to the project? WIthout making the code unapproachable for new contributors? Can they be optional?

A common way to introduce lenses is to add them in a completely new module like `Hledger.Lenses`. Or it can be introduced by prepending "L" to all of your record methods and having them in the same module as the original record. Also as an added benefit, it is possible to write lenses without importing the library, like this:

```haskell

type Lens s t a b = forall f. Functor f => (a -> f b) -> s -> f t
type Lens' s a = Lens s s a a

jtxns :: Lens' HL.Journal [ HL.Transaction ]
jtxns f s = (\b -> s { HL.jtxns = b }) <$> f (HL.jtxns s)
```

Doing so will have the minimum impact on users down the line, as it requires no change of code and the project still compiles with dependencies. 


> This nicely solves a problem I was having (how to look upward in the account hierarchy to generate Ledger's compressed-tree-style balance reports)?

Can you point me to the point in the code? Then I can see if I can come up with a solution that does not use the recursive data structure.

Thanks again, 
Christian

**Comment by simonmichael** [comment]
It was particularly to allow walking up the account tree, as well as down, to calculate the boring parents to elide (squash onto one line) in the single column tree-mode balance report. 

Today's code looks different, but I think it's used in https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Reports/BalanceReport.hs#L98-L138, probably line 132 calling parentAccounts function which relies on those upward references. Unless I'm confused.

**Comment by kalhauge** [comment]
Okay, I have been playing a little around with trying to untie the account tree. 

This is not a pull request (https://github.com/kalhauge/hledger/commit/0a0969b27dc0924c9b79e315ea4d1b313d144a7c) but more an idea about how to calculate the boring parents without using recursive types. When I worked on removing the parent field, I could see Account and the Ledger Type is only used as intermediate results in the balance calculation. So I removed them and did the calculations using data structures instead. The commit passes *most* of the test cases. 

The basic Idea is to collect the accounts in two different data-structures depending on whether the tree should be flat or not. 

The flat version just used the Monoidal map from before and is straightforward
(https://github.com/kalhauge/hledger/blob/0a0969b27dc0924c9b79e315ea4d1b313d144a7c/hledger-lib/Hledger/Reports/BalanceReport.hs#L177):
```haskell
    flatReport = (items, total)
      where
        balance = journalExclusiveBalance q j
        total = getSum . Map.foldMapWithKey (\_ s -> s) . unMonoidalMap $ balance
        items =
          map (\(name, total) -> (aname' name, aname' name, 0, getSum total))
          . Map.toAscList
          . unMonoidalMap
          $ balance
```

The tree version is more complicated. There I use a ad-hoc data-stucture, which I call a Trie. The trie makes sure to add its value to all parents of an Account. Here I uses a `type Account' = [AccountName]`. 

Here I reduce the trie (https://github.com/kalhauge/hledger/blob/0a0969b27dc0924c9b79e315ea4d1b313d144a7c/hledger-lib/Hledger/Reports/BalanceReport.hs#L187):
 
```haskell
    treeReport =  snd $ trieReduce reductor balance
      where
        balance = journalInclusiveBalance q j

        reductor
          :: Sum MixedAmount
          -> [(T.Text, (Account', BalanceReport))]
          -> (Account', BalanceReport)
        reductor (Sum total) subs = traceShowId $
          case subs of
            -- An item is boring if it only has one sub,
            -- if that is the case promote that sub to be the active node.
            [(name, (an, (items', total')))]
              | total' == total ->
              (name:an, (map (addName name) items', total))
            _ ->
              -- If more or zero subs combine them, an choose the empty
              -- name space
              ([], (concatMap (uncurry combine) subs, total))

        combine name (an, (items', total'))
          | isZeroMixedAmount total' && null items' && not (empty_ opts) = []
          | otherwise =
              (aname' (name:an), aname' (name:an), 0, total')
              : map (addName name) items'

        addName name (an, elied, i, ma) =
          (aname' [name, an], elied, i + 1, ma)
```

There is still a lot of work to be done, and these changes are *clearly* breaking in nature. This means that it should probably (if at all) be done as a massive pass to clean up the entire API.

What do you think?

Some Notes:

* I had to comment out hledger-ui, because hsfevents does not compile on OSX/Stack/Nix. Maybe splitting hledger-lib and the others into self-contained repositories might reduce building overhead.
* Is It correct to let the query be depended on the report options. (https://github.com/simonmichael/hledger/blob/5c85a1dd1c5348c98eef0674378dff619dd211c4/hledger-lib/Hledger/Reports/ReportOptions.hs#L340) This is the main reason, that some of the test cases fail.

**Comment by simonmichael** [comment]
Great start, thanks for exploring this. 

I haven't had a chance to read that giant commit yet. Does it include unrelated changes ? I can review single-purpose commits much quicker.

It sounds like a lot of work, and we have existing working battle-tested code. This will have to show clear benefits, without degradation in metrics such as memory allocation, maximum residency, and run time.

Can you open a separate issue about your hledger-ui problem - it is supposed to build on OSX.

I didn't totally understand your last question - I think the answer is yes, hopefully looking at the existing code will resolve it.

**Comment by simonmichael** [comment]
PS, some more context: there are too many variants of *balanceReport. Of these, I believe only the original balanceReport has the boring parent eliding functionality which depends on the mutual referencing IIRC. If/when we can implement this for single/multiBalanceReport, the old balanceReport can be dropped. (Maybe you've already done this work, not sure yet.)

Is your adhoc structure actually a Trie ? (I ask because of "...I call it a Trie...")

**Comment by kalhauge** [comment]
Thanks for bearing with me!

>I haven't had a chance to read that giant commit yet. Does it include unrelated changes? I can 
> review single-purpose commits much quicker.

Sorry, It's so long, but it was the shortest commit that still compiled and didn't fail any of the test cases. I'm mostly removing files so the most interesting part of the commit is in
hledger-lib/Hledger/Reports/BalanceReport.hs: 

https://github.com/kalhauge/hledger/commit/0a0969b27dc0924c9b79e315ea4d1b313d144a7c#diff-0ad5f40b03b67d6304541a5b62d3ff18

>It sounds like a lot of work, and we have existing working battle-tested code. This will have to 
> show clear benefits, without degradation in metrics such as memory allocation, maximum 
> residency, and runtime.

This is a lot of work! And it will break a lot of code! So you are completely right to have some reservations :). I can't promise that it would be worth the time or improve the metrics. But here are some predictions:

- Removing tieing will have some performance improvements. Tieing essentially creates a copy of all the records it changes. This means that you have to iterate through the entire ledger file at least twice. Once for reading, and once for tieing. 

- Tieing also removes the possibility to do lazy loading. You have to read the entire file before you can tie the knot on the transactions. Haskell is sometimes able to do operations over streams that keep the operations in constants space. Since most of the operations in hledger are read and throw away it might improve garbage collection time.

- Doing things when needed and not upfront is often preferable. This means that calculating all the information that is calculated in the Account type might not be necessary for the report. Example the aibalance and aebalance are not needed at the same time, but they both need to be calculated if we want to create an Account.

> I didn't totally understand your last question - I think the answer is yes, hopefully looking at the existing code will resolve it.

I think that I wanted to point out that it makes sense (to me) to separate query and report options. In my mind, the query selects the things that you want to display and the report options describe how to display it. I especially reference to these test-cases: 

https://github.com/kalhauge/hledger/blob/5c85a1dd1c5348c98eef0674378dff619dd211c4/hledger-lib/Hledger/Reports/BalanceReport.hs#L230-L244

Here both depth:2 and --depth=2 yields the same results, which also means that you cannot query the ledger on depth, ei.  You can't say that you want to sum all amounts to accounts, not counting accounts of depth more than 2.

> Can you open a separate issue about your hledger-ui problem - it is supposed to build on OSX.

It's not really a problem with heldger-ui, its a problem with hsfevents. But I can open a bug report regardless.

>  If/when we can implement this for single/multiBalanceReport, the old balanceReport can be dropped. (Maybe you've already done this work, not sure yet.)

I only did the simple BalanceReport.

> Is your adhoc structure actually a Trie ? (I ask because of "...I call it a Trie...")

I think it is actually a generalization of a Trie: https://en.wikipedia.org/wiki/Trie. In my trie, the value is any monoid, so the trie represented in the article could be represented like:

```haskell
WikiTrie a =~ Trie Char (First a)
```

**Comment by simonmichael** [comment]
Status-sm: waiting for spare cycles...

**Comment by simonmichael** [comment]
@kalhauge, do you have it working without knot-tying/circular references ? If so, how has the memory usage/run time been affected ?

I may have added confusion in past discussions, thinking there was only one use of circular references in hledger. I believe there are two: Transaction <-> Posting, and parent Account <-> child Account. I think you are working only on removing the Account circularity, so far.

**Comment by kalhauge** [comment]
@simonmichael, sadly I suddenly got very busy, so I haven't had any time to work on the project since the last commit. I will try to get around to it soon. Do you have a standard way of testing memory usage and run time? 

I did notice the Transaction <-> Posting circular reference, but I wanted to take a smaller example first.

**Comment by simonmichael** [comment]
One way:

`$ stack build hledger`
`$ stack exec -- hledger +RTS -s -RTS -f examples/1000x1000x10.journal bal >/dev/null`
Look at maximum residency and total time elapsed.

**Comment by simonmichael** [comment]
Also, to make the sample journals:

`$ make samplejournals`

Other commands:

`$ stack exec -- time hledger -f examples/1000x1000x10.journal bal >/dev/null`
`$ stack install quickbench`
`$ stack exec -- quickbench "hledger -f examples/1000x1000x10.journal bal >/dev/null"`

**Comment by kalhauge** [comment]
It took a little work to get the new version to work.

My version is using more garbage collection but I haven't spend a lot of time optimizing 
yet. 

https://github.com/kalhauge/hledger/commit/e408bbb773bc2990e7c3dbc27eb05bb01aac55fb

This is the new version:
```
$ stack exec -- hledger +RTS -s -h -p -RTS -f examples/1000x1000x10.journal bal >/dev/null
   1,095,250,368 bytes allocated in the heap
      79,688,672 bytes copied during GC
       6,275,920 bytes maximum residency (15 sample(s))
          56,496 bytes maximum slop
              21 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1046 colls,     0 par    0.143s   0.211s     0.0002s    0.0507s
  Gen  1        15 colls,     0 par    0.001s   0.002s     0.0001s    0.0003s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.005s elapsed)
  MUT     time    1.100s  (  1.218s elapsed)
  GC      time    0.122s  (  0.182s elapsed)
  RP      time    0.000s  (  0.000s elapsed)
  PROF    time    0.023s  (  0.031s elapsed)
  EXIT    time    0.000s  (  0.002s elapsed)
  Total   time    1.247s  (  1.407s elapsed)

  Alloc rate    995,292,179 bytes per MUT second

  Productivity  88.3% of total user, 84.5% of total elapsed
```

So this is the old version:

```
$ stack exec -- hledger +RTS -s -h -p -RTS -f examples/1000x1000x10.journal bal >/dev/null
   1,057,191,344 bytes allocated in the heap
      76,262,504 bytes copied during GC
       6,173,664 bytes maximum residency (14 sample(s))
          56,504 bytes maximum slop
              19 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1012 colls,     0 par    0.149s   0.229s     0.0002s    0.0653s
  Gen  1        14 colls,     0 par    0.002s   0.002s     0.0002s    0.0009s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.011s elapsed)
  MUT     time    1.190s  (  1.273s elapsed)
  GC      time    0.125s  (  0.195s elapsed)
  RP      time    0.000s  (  0.000s elapsed)
  PROF    time    0.026s  (  0.036s elapsed)
  EXIT    time    0.000s  (  0.002s elapsed)
  Total   time    1.343s  (  1.481s elapsed)

  Alloc rate    888,437,896 bytes per MUT second

  Productivity  88.6% of total user, 83.6% of total elapsed
```

**Comment by simonmichael** [comment]
Excellent data!

So, no significant change yet. 

By the way, if you're not already an expert on strictness I can highly recommend FP Complete's [All About Strictness](https://www.fpcomplete.com/blog/2017/09/all-about-strictness) (I just read it). 

I imagine that current hledger code will have not one or two, but many overlapping causes of space inefficiency. I expect it will take considerable perseverance in understanding and cleaning these before we start seeing significant gains.

**Comment by kalhauge** [comment]
Thanks for the article it looks really cool!

What would be the plan going forward?

**Comment by simonmichael** [comment]
Whaddya mean, I though you had the plan!

Actually I think I might have confused this issue with another which claimed to improve performance. 

So, what plan are we talking about here.. for which goal ?

**Comment by kalhauge** [comment]
I don't want to take over your project, but I would love to help you untieing the
data structures and improving the API so that It is easier to work with and potentially faster.
Short list of Ideas:

- Clean up Transactions and Postings
- Remove Types.hs completely, distribute data declarations to the corresponding 
  Modules.
- Change the Journal to be a Stream instead of a Data Structure. Most operations are 
  single pass, which means that streaming the contents from a file will result in constant 
  space instead of linear.

I would love to help with these tasks, but only if you think it would help :).

**Comment by simonmichael** [comment]
Excellent. I am interested in things that produce real benefits for newcomers, regular users, new contributors, regular developers, packagers, documentors, and/or maintainers. What do you think might be a small and fruitful direction to explore ?

**Comment by simonmichael** [comment]
My 2 cents: in the area of things we've been talking about, these would be quite valuable:

- add a new focus on performance to the project, by making it routine and part of our workflow to measure it and track it over time. There should be a single metric that we keep in mind and try to improve, I'm thinking maximum memory residency. We use too much space and reducing that will also reduce time.

- review past work on improving memory usage. There is one issue I'm thinking of with graphs and the platform:windows label.

- revive/set up the workflows for generating a standard time & space profile, and maybe a graphical heap profile, with latest tools (GHC 8.2.1 etc.)

- study the new profiles and identify some likely low-hanging fruit for optimisation

**Comment by simonmichael** [comment]
PS none of the above are the tasks you mentioned, but I think they are prerequisites to merging such deep changes, so they would enable those kinds of cleanup.

**Comment by kalhauge** [comment]
I totally agree. Regular benchmarking should be prioritized.

Maybe look at this framework. It is very robust, and as I can tell the industry standard.
https://hackage.haskell.org/package/bench

It is also possible to integrate with stack.

**Comment by simonmichael** [comment]
I know it; I wrote https://hackage.haskell.org/package/quickbench to get output I could read more easily. In the makefile and maybe contributors guide there is some benchmarking.

**Comment by kalhauge** [comment]
After some research, I think that one approach would be to: 

1) Write the benchmarks using [Critierion](http://www.serpentine.com/criterion/tutorial.html)
2) Run the benchmarks for each comit using [feed-gipeda](https://github.com/sgraf812/feed-gipeda) or [colben](https://hackage.haskell.org/package/cloben).
3) Display them using [Gipeda](https://hackage.haskell.org/package/gipeda)

Maybe this can be integrated into a Jenkins server?

**Comment by simonmichael** [comment]
[Here](https://github.com/simonmichael/hledger/blob/master/hledger/hledger.cabal#L251) is our current most formal benchmark. It runs [bench.hs](https://github.com/simonmichael/hledger/blob/master/hledger/bench/bench.hs), which does runs several operations on a 10000-transaction journal. It does not use Criterion by default; for that you must run it yourself with `--criterion`. (Hint: you'll get the same results, but it will take much longer.) Currently nobody sees this benchmark and it has no real purpose.

In the past I sometimes included quick end-user-relevant benchmark results in release notes, like [here](http://hledger.org/release-notes.html#hledger-0.24). More recently I just do it occasionally during development, with `make quickbench`. This kind of report uses the quickbench tool, which was released separately. 

See `make` -> "TESTING:" for more performance-related rules in the makefile.

**Comment by kalhauge** [comment]
I find benchmarking very interesting. I'm currently working on a side project that also has a performance angle to it. I feel that there are two angles to it. 

Publication benchmarking, which primary goal is to ensure your users that the program is performing better over time. This is where tools like "Gipeda" or "colben" are great as they
are external to the day to day development. 

Regression style benchmarking, where we want to ensure that our code is not getting slower. These benchmarks should be run every commit before committing, in the same way as unit test should be run before each commit. The idea being that you should not commit slower code, the same way as you should not commit code with failing tests.

But, maybe we should have this conversation in another issue/place?

**Comment by simonmichael** [comment]
Yes, let's discuss further on the mail list or IRC channel.

**Comment by simonmichael** [comment]
To be continued in discussion forums and specific issues.

**Comment by kalhauge** [comment]
Agree!
